### PR TITLE
ci: harden the shared release-gate scripts and llm-smoke roster golden

### LINKED ISSUE #190: Follow-up hardening for the shared release-gate scripts and llm-smoke roster golden
Follow-up hardening from the LLM smoke publish-gate work (#154). None of these are correctness bugs in the shipped gate; they are coverage and hygiene gaps found while building it. Several are pre-existing in the connector gate and were inherited unchanged by the now-shared scripts, so they are worth fixing against the shared code rather than either gate alone.

## Test coverage

- **The roster golden does not pin the fan-in wiring.** `test/llm-smoke-roster.unit.test.sh` parses only the `- leg:` matrix rows, so it validates the six legs but not the `ROSTER`, `JOBS`, and `PROOFS` literals in `gate-assert`. Dropping a leg from `ROSTER` (for example `bedrock-tools:aws-oidc`) leaves the golden green. The backstop today is that a genuinely failing leg still fails its own job and blocks `gate_sha` via the result check, so this is not a live fail-open, but a leg silently removed from the roster would go ungated. Extend the golden to pin the three literals against the canonical roster it already anchors the matrix rows to.
- **Assert the two smoke env guards structurally.** Add a check that each `llm-smoke)` dispatch branch in the four credential jobs exports `SMOKE_REQUIRE_NO_CONNECTORS=1`, and that `SMOKE_REQUIRE_RUN: "1"` is present in every leg's `env:`. The latter is the more consequential and is defended today only by a hardcoded literal plus a comment.
- **Three `ci-gate-assert.sh` guards have no covering test** (found by mutation testing): the duplicate-job guard's cross-family shape, an inactive job present in `JOBS` and `needs` but absent from `RESULTS`, and the two halves of the `NEEDS_JSON` type guard. Add a case for each.
- No test pins a truly unset `DISPATCH_POLICY` in `ci-gate-contract.sh` (empty is covered; unset is verified safe by hand). Cheap case to add.

## Shared-script hygiene

Inputs to both scripts are static literals authored in-workflow, not attacker-controlled, so these are robustness and consistency fixes, not exposures:

- Add `set -f` around the `for _c in all $SELECTORS` glob in `ci-gate-contract.sh` (pre-existing in the connector gate), and extend `set -f` to the other unquoted-split loops in `ci-gate-assert.sh`.
- Add an arity check so a malformed `JOBS` entry like `a:b:c:always` cannot silently swallow a field.
- Drop the dead `family=` store in `ci-gate-assert.sh`'s inactive-job loop.
- Decide whether a dispatch `SELECTOR` that names no leg, with all jobs skipped, should exit 0 or fail closed (pre-existing; only reachable via a filtered dispatch, and the release path always passes an empty selector).

## Docs and comments

- `ci-gate-contract.sh` and its unit test still carry "connector" wording in a now-shared context; the `ci-gate-assert.sh` family-coupling comment implies both directions are load-bearing when only `JOBS -> ROSTER` is. Reword for the shared scope.
- Soften the roster golden's header comment so its stated scope matches what the parser actually checks.

## Pre-existing, low urgency

- In `release.yaml`, the `publish` gate's `gate_sha == needs.release.outputs.sha` comparison would be satisfied by `'' == ''` if the release SHA were ever empty while `release_created == 'true'`. Not reachable in practice (`ensure-tag.sh` fails earlier), and shared with the connector gate. An explicit non-empty assertion would close it.
- Consider a Makefile grep-pin for the publish gate condition alongside the existing `GATE_WORKFLOWS`/`EXPECTED_CALLER` check, covering both the connector-e2e and llm-smoke clauses.

Refs #154.

### PR DESCRIPTION
**Pull request by 0xh0b0**
## What & why

Follow-up hardening from the LLM smoke publish-gate work, closing the coverage and hygiene gaps in the shared release-gate scripts. None of these were live fail-opens; they are the guards that keep it that way. Closes #190.

Shared scripts:
- `ci-gate-assert.sh`: script-wide `set -f` (every word list is deliberately split unquoted, so a glob character in a token can no longer expand against the working directory), exact-reconstruction arity and non-empty checks on `JOBS` triples and `ROSTER` pairs (a malformed entry like `a:b:c:always` used to silently drop a field), a fail-closed check that a dispatch selector names at least one roster leg (previously a drifted selector exited 0 claiming the leg "ran and passed" while nothing ran), and a dead store removed. The family-coupling comment now says which direction is load-bearing.
- `ci-gate-contract.sh`: script-wide `set -f` and shared-scope wording (the script serves both gates now, and the "connector selector" die message predated that).

Test coverage:
- The gate-assert unit suite gains an optional stderr-diagnostic assertion in its harness, plus cases for the previously untested guards: the duplicate-job guard across families, an inactive job missing from `RESULTS`, both halves of the `NEEDS_JSON` type guard (the array-of-job-names case is the one only `isinstance` can catch; the `{}` case pins the guard's own diagnostic since its exit code is reachable downstream), the new arity guards, the no-match selector, and a distinguishing glob fixture that flips outcome on `set -f` itself.
- The contract suite now exercises a truly unset `DISPATCH_POLICY` (only empty was covered).
- The roster golden now also derives the gate-assert `ROSTER`/`JOBS`/`PROOFS` literals from the same canonical block that anchors the matrix rows (a leg dropped from the fan-in is invisible to the runtime checks because the remaining legs still pass), and pins each smoke step's operational seam: the `SUITE`/provider/model matrix bindings and the full suite dispatcher, so the connector-dark tripwire cannot be disabled while the rows stay green.

Publish gate:
- `release.yaml`: an explicit `needs.release.outputs.sha != ''` conjunct, closing the degenerate `'' == ''` satisfaction of the two `gate_sha` equality checks.
- `Makefile`: `sh-test` now pins the publish gate's five required conjuncts against the extracted live `publish` job `if:` line (never a whole-file grep, which a comment or dead job could satisfy), alongside the existing trusted-caller pin.

## Checklist
- [x] `make check` passes locally
- [x] Tests added/updated for the change
- [x] Docs updated if behavior changed

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -166,7 +166,29 @@ sh-test:
 			exit 1; \
 		fi; \
 	done
-	@echo "OK: sh-test (install.sh unit + loopback smoke + e2e guardrails unit + connector-e2e unit + render-scoop unit + dependabot-override unit + ci-gate-contract unit + ci-gate-assert unit + llm-smoke roster unit + python syntax gate + connector audit parsers + shared-machinery selftest + connector-e2e trusted-caller pin check)"
+	@# The publish gate's compound if is the sole consumer of the two gate proofs; a
+	@# term dropped there (or a gate_sha comparison edited away) would let a red or
+	@# absent gate publish anyway, and nothing else would notice before a live
+	@# release. Scoped to the publish job's own if: line, never a whole-file grep,
+	@# which a comment or a dead job could satisfy.
+	@publish_if=$$(awk '/^  publish:$$/{injob=1} injob && /^    if: /{print; exit}' .github/workflows/release.yaml); \
```
[the remark this answers is not in the document]
Comment (0xh0b0) [reply Makefile]: Fixed in 5ba21a4: the scan now resets at every job header, so a publish job missing its if: hits the locate failure instead of latching onto a later job's line.

File: test/llm-smoke-roster.unit.test.sh [hunk 2]
```diff
@@ -117,6 +122,193 @@ while i < len(lines):
...
+    "gcp-wif": "always",
+    "aws-oidc": "always",
+    "api-key-release": "release",
+    "api-key-manual": "manual",
+}
+
+canonical = []
+for raw in open(sys.argv[2]).read().splitlines():
+    if raw.strip():
+        canonical.append(raw.split("|"))
```
[the remark this answers is not in the document]
Comment (0xh0b0) [reply test/llm-smoke-roster.unit.test.sh]: Applied in b63d916, to both open() calls so the parser stays consistent.
